src/deploy/functions/containerCleaner.ts

```diff
@@ -360,20 +360,24 @@ export interface Stat {
 
 export class DockerHelper {
   readonly client: docker.Client;
-  readonly cache: Record<string, Stat> = {};
+  readonly cache: Record<string, Promise<Stat>> = {};
 
   constructor(origin: string) {
     this.client = new docker.Client(origin);
   }
 
+  // N.B. It is very important that this function assigns to the cache before
+  // yielding the runloop or parallel executions will race their LS calls, which
+  // are each recursive, leading to possible N^2 executions.
   async ls(path: string): Promise<Stat> {
-    if (!this.cache[path]) {
-      const raw = await retry(() => this.client.listTags(path));
-      this.cache[path] = {
-        tags: raw.tags,
-        digests: Object.keys(raw.manifest),
-        children: raw.child,
-      };
+    if (!(path in this.cache)) {
+      this.cache[path] = retry(() => this.client.listTags(path)).then((res) => {
+        return {
+          tags: res.tags,
+          digests: Object.keys(res.manifest),
+          children: res.child,
+        };
+      });
```

Comment:
Sorry to bother. I think using `cache = Object.create(null)` is safer. What if `path` here is `toString`?

Reply:
That would also work, but I'm not sure what you mean with "`path` here is `toString`"! Can you elaborate?

Reply:
Sure. pass `toString` as the argument to the method `ls(path)`.

Reply:
It seems a little overboard, but sure, why not.